### PR TITLE
[python] convert datatable to numpy directly

### LINKED ISSUE #1968: memory leak with H2O's DataTable
[creep.zip](https://github.com/Microsoft/LightGBM/files/2794219/creep.zip)

When doing a permutation variable importance one loops over many shuffled columns and get predictions over and over.

We have seen that LightGBM has a memory leak when performing this operation.  XGBoost has no memory leak.

Just run the script and keep track of (say) resident memory for the process.  

It starts out bouncing around 0.4GB, but by n=1300 columns later bounces around 1.0GB.  Once it hit n=2100, it's bouncing around 1.6GB.  By n=2500 columns, memory bouncing around 1.8GB.

You can change get_new_preds to False to just see how datatable is not using any extra memory.  Go back and then try xgboost and no such memory leak occurs.

Here is the location of the data, and one needs to fix the path used in the code: https://www.kaggle.com/c/santander-value-prediction-challenge/download/train.csv (you'll need a kaggle account).

Thanks!
Jon

### PR DESCRIPTION
**Pull request by guolinke**
fix #1968

### CODE REVIEW

**Comment by guolinke** [comment]
Thanks @StrikerRUS , I push some fixes.

File: python-package/lightgbm/basic.py [hunk 1]
```diff
@@ -1398,6 +1402,8 @@ def get_data(self):
                 self.data = self.data[self.used_indices, :]
             elif isinstance(self.data, DataFrame):
                 self.data = self.data.iloc[self.used_indices].copy()
+            elif isinstance(self.data, DataTable):
+                self.data = self.data[self.used_indices, :]
```
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply python-package/lightgbm/basic.py]: Me too... To be honest, I see DataTable for the first time 😄 

Speaking about pandas, I used `.copy()` with the aim to prevent [`SettingWithCopyWarning`](https://www.dataquest.io/blog/settingwithcopywarning/) in users' preprocess functions.

File: python-package/lightgbm/sklearn.py [hunk 2]
```diff
@@ -479,7 +479,7 @@ def fit(self, X, y,
             eval_metric = [eval_metric] if isinstance(eval_metric, (string_type, type(None))) else eval_metric
             params['metric'] = set(original_metric + eval_metric)
 
-        if not isinstance(X, DataFrame):
+        if not isinstance(X, DataFrame) and not isinstance(X, DataTable):
```
Comment (StrikerRUS) [inline python-package/lightgbm/sklearn.py]: Just letting you know that it can be `if not isinstance(X, (DataFrame, DataTable)):` for short.

**Comment by StrikerRUS** [comment]
I've updated more docstrings and simplified `isinstance` check.

File: python-package/lightgbm/basic.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (guolinke) [reply python-package/lightgbm/basic.py]: Okay, I see. So theoretically, we don't need a copy here.
@pseudotensor any idea about this?

File: python-package/lightgbm/compat.py [hunk 3]
```diff
@@ -90,6 +90,19 @@ class DataFrame(object):
 except ImportError:
     GRAPHVIZ_INSTALLED = False
 
+"""datatable"""
+try:
+    from datatable import DataTable
+    DT_INSTALLED = True
```
Comment (StrikerRUS) [inline python-package/lightgbm/compat.py]: @guolinke Don't you mind to rename this variable to `DATATABLE_INSTALLED`, for the consistency with other variables (for example, there are `PANDAS_INSTALLED` but not `PD_INSTALLED`). Also, `DT` is a little bit confusing: sometimes `dt` is used for `datetime`.

**Comment by guolinke** [comment]
@StrikerRUS I think we can merge this, how do you think?

**Comment by StrikerRUS** [comment]
@guolinke I'm just worrying about the need of `copy`...

File: python-package/lightgbm/basic.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply python-package/lightgbm/basic.py]: ping @pseudotensor 
Please take a look at this function for getting raw data.

**Comment by guolinke** [comment]
@StrikerRUS I searched 'copy' in datatable documents, and didn't find anything.
https://datatable.readthedocs.io/en/latest/search.html?q=copy&check_keywords=yes&area=default
I think they don't need this.

We also can merge this first, and provide the fix for that in the future if needed.

**Comment by StrikerRUS** [comment]
@guolinke Agree!

Seems that `numpy` and `DataTable` slicing cases can be merged, if we don't need to `copy`:
```
if isinstance(self.data, (np.ndarray, DataTable)) or scipy.sparse.issparse(self.data):
    self.data = self.data[self.used_indices, :]
```

**Comment by guolinke** [comment]
I prefer to separate them, for the potential future needs.
